README
Ginger
======

Ginger is an open source host management plugin to Wok, Web Server Originated
from Kimchi, that provides an intuitive web panel with common tools for
configuring and operating Linux systems.

Wok is a cherrypy-based web framework with HTML5 support that is extended by
plugins which expose functionality through REST APIs.

The current features of Host Management of Linux system include:
 + retrieve system health (sensors) stats
 + user login account management,
 + network interface configuration,
 + configuration backup,
 + Power (ppc) firmware update,
 + Power policy management.

Browser Support
===============

Wok and its plugin can run in any web browser that supports HTML5. The
Kimchi community (responsible for Wok project) makes an effort to
test it with the latest versions of Chrome and Firefox browsers, but the
following list can be used as reference to browser support.

Desktop Browser Support:
-----------------------
* **Internet Explorer:** Current version
* **Chrome:** Current version
* **Firefox:** Current version
* **Safari:** Current version
* **Opera:** Current version

Mobile Browser Support:
-----------------------
* **Safari iOS:** Current version
* **Android Browser** Current version


Hypervisor Distro Support
=========================

Kimchi might run on any GNU/Linux distribution that meets the conditions
described on the 'Getting Started' section below.

The Kimchi community makes an effort to test it with the latest versions of
Fedora, RHEL, IBM KVM and Ubuntu.

Getting Started
===============

Install Dependencies
--------------------

**For fedora, RHEL and IBM KVM :**

    $ sudo yum install  libvirt-python libvirt hddtemp \
                        libuser-python python-augeas python-netaddr\
                        python-ethtool python-ipaddr python-magic \
                        pyparted gettext-devel \

    # For fedora, RHEL and Power ppc
    $ sudo yum install tuned lm_sensors

    # Power ppc install the following packages
    $ sudo yum install powerpc-utils serviceable.event.provider

    # These dependencies are only required if you want to run the tests:
    $ sudo yum install pyflakes python-pep8 python-mock

    # For UI development
    $ sudo yum install gcc-c++ python-devel python pip
    $ sudo pip install cython libsass


**For openSUSE:**

    # Add repository for hddtemp:
    $ sudo zypper ar -f http://download.opensuse.org/repositories/utilities/openSUSE_Leap_42.1/ utilities

    # Add repository for python-parted:
    $ sudo zypper ar -f http://download.opensuse.org/repositories/home:/GRNET:/synnefo/openSUSE_Leap_42.1/ home_GRNET_synnefo

    # Add repository for python-ethtool:
    $ sudo zypper ar -f http://download.opensuse.org/repositories/devel:/tools/openSUSE_Leap_42.1/ devel_tools

    # Add repository for python-magic:
    $ sudo zypper ar -f http://download.opensuse.org/repositories/home:/Simmphonie:/python/openSUSE_Leap_42.1/ home_Symmphonie_python

    $ sudo zypper install libvirt-python libvirt hddtemp \
                        libuser-python python-augeas python-netaddr\
                        python-ethtool python-ipaddr python-magic \
                        python-parted gettext-tools \

    # These dependencies are only required if you want to run the tests:
    $ sudo zypper install python-pyflakes python-pep8 python-requests python-mock

    # For UI development
    $ sudo zypper install python-devel python pip
    $ sudo pip install cython libsass


**For Ubuntu (Debian-based):**

    $ sudo apt-get install libvirt-bin python-libvirt hddtemp \
                        python-libuser python-ethtool python-augeas \
                        python-ipaddr python-magic python-netaddr \
                        python-parted gettext \

    # These dependencies are only required if you want to run the tests:
    $ sudo apt-get install pep8 pyflakes python-requests python-mock

    # For UI development
    $ sudo apt-get install g++ python-dev python pip
    $ sudo pip install cython libsass

At the moment of this writing (11-30-2015), libuser has a bug that
prevents it to work properly in Ubuntu 15.10 and older
(https://bugs.launchpad.net/ubuntu/+source/libuser/+bug/1387274).
A workaround is needed to fix this bug:

$ sudo touch /etc/libuser.conf

This workaround is for Debian/Ubuntu only.

Build and Install
-----------------

    Wok Framework Installation:
    $ ./autogen.sh --system

    $ make
    $ sudo make install   # Optional if running from the source tree

    Ginger Base:
    $ cd src/wok/plugins/gingerbase
    $ ./autogen.sh --system

    $ make
    $ sudo make install   # Optional if running from the source tree

    Ginger:
    $ cd ginger
    $ ./autogen.sh --system

    $ make
    $ sudo make install   # Optional if running from the source tree

Run
---
    $ sudo /usr/bin/wok --host=0.0.0.0

    OR

    $ sudo systemctl start wokd

*** Notes on Power policy management feature ***

The power policy management feature uses the 'tuned' service to control the
power policies of the host. Problems have been reported with this package,
depending on the host configuration, such as SELinux denials and trouble
to communicate using DBUS with the 'tuned' service started from systemd.

If you find any problems with the power policy management feature, we
recommend following these steps (all steps requires 'sudo' privileges):

- put SELinux in permissive mode for 'tuned' (required if the version of the
package 'selinux-policy' is < 3.11):

$ semanage permissive -a tuned_t

- disable the 'tuned' service from systemd and restart it by hand:

$ systemctl stop tuned.service
$ tuned -l -P -d

Remember to restart wokd service after these changes:

$ systemctl restart wokd.service

If these steps do not solve the problem, try to update all 'tuned' related
packages and the packages'selinux-policy' and 'selinux-policy-targeted'.


UI Development
----
Make sure to update the CSS files when modifying the SCSS files by running:

    $ sudo make -C ui/css css


Test
----

    $ cd ginger
    $ make check-local # check for i18n and formatting errors
    $ sudo make check

After all tests are executed, a summary will be displayed containing any
errors/failures which might have occurred.

Participating
-------------

All patches are sent through our mailing list.  More information can be found at:

https://github.com/kimchi-project/ginger/wiki/Communications

Patches should be sent using git-send-email to ginger-dev-list@googlegroups.com

** Copyright notice ***

All the .gif and .png files in 'ui/css/base/images/' directory are licensed
as follows:

----------------------------
Copyright IBM Corp, 2015

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA


ChangeLog
CHANGELOG
=========

#### [2.4.0] ####
 * [e5b9422] Updating PO files                                                              (Daniel Henrique Barboza)
 * [503be72] Move System Modules to a sidebar entry                                         (Aline Manera)
 * [ffca4f7] Move Audit features to a sidebar entry                                         (Aline Manera)
 * [33a343c] Ginger sidebar                                                                 (Aline Manera)
 * [b3bb340] Fix expand/collpase widget configuration in Network tab                        (Aline Manera)
 * [aedb1e5] Updated PO files                                                               (Rajat Gupta)
 * [b3510f0] Fix mocking of test_get_list_of_audit_conf                                     (Lucio Correia)
 * [9551e0a] Skip tests that need sudo                                                      (Lucio Correia)
 * [744e29c] Fix and mock test_update for Audit rules                                       (Lucio Correia)
 * [044a8f4] Fix and improve test_get_list_of_pre_rules                                     (Lucio Correia)
 * [16b6a20] Adapt filesystem tests to also run without sudo permission                     (Lucio Correia)
 * [bcbd516] Bug fix #378: Change confirmation dialog of password mismatch in Add User wi.. (Bianca Carvalho)
 * [cb20d1a] Fixing copyright date after previous commits                                   (Daniel Henrique Barboza)
 * [f406292] Externalized active value in audit dispatcher details and list sections.       (Rajat Gupta)
 * [ff80f99] Fixed duplication of user list in user management user list                    (Rajat Gupta)
 * [f61d64d] Introducing Volume group action buttons                                        (Rajat Gupta)
 * [e89a82c] Added validation and fixed error message while adding swap device              (Rajat Gupta)
 * [c29973c] Fixed volume group listing will not refresh after extend/reduce volume         (Rajat Gupta)
 * [34e32f8] Fixed Apply button of audit config can be clicked when no changes was made t.. (Rajat Gupta)
 * [b5afa32] Issue #527.Adding /dev/dasd based swap device yields an error.                 (Megha Smriti)
 * [9560100] Fixing copyright date of UI files                                              (Daniel Henrique Barboza)
 * [68c6ecf] Changes to show translated value of rules types in rules listing and edit ru.. (Rajat Gupta)
 * [bc9b25d] Externalise Network on-off button for ipv4/ipv6 panels and Host Administrati.. (rajat gupta)
 * [7296802] Multi culture UI issues fixed                                                  (Rajat Gupta)
 * [6721ad8] Collapse all Network tab sections                                              (Aline Manera)
 * [f0f20b3] Issue #520. Auto refresh of ovsbridges.                                        (Megha Smriti)
 * [57b0e8f] Update run_server() calls to do not pass model instance                        (Aline Manera)
 * [b203d32] Fixing test_interfaces.py                                                      (Daniel Henrique Barboza)
 * [30e65e7] disable remove button for iscsi devices                                        (Suresh Babu Angadi)
 * [2536442] Fixing copyright of host-network-ovs.js and host-system-services.js            (Daniel Henrique Barboza)
 * [bbddec8] Fixed system service tab missed to list all active services                    (Rajat Gupta)
 * [186a740] Fixed ovs bond interface menu will not exit when save changes and close "Edi.. (Rajat Gupta)
 * [84872f7] Fix for issue #529 and #530                                                    (Suresh Babu Angadi)
 * [10be1fa] Adding self.depends in ginger root                                             (Daniel Henrique Barboza)
 * [f283380] Fixing copyright of ui/js/host-storage-vg.js                                   (Daniel Henrique Barboza)
 * [c2d2bd0] Fixed GINVG00001E and GINVG00009E error received while creating a new volume.. (Rajat Gupta)
 * [1cf1359] Fixed missing apply step when try to extend volume group                       (Rajat Gupta)
 * [ce2b385] Fixed window hangs at the 'loading' step when cancel the physical volume cre.. (Rajat Gupta)
 * [df814a0] Corrected wrong spelling of warning info while creating physical volume.       (Rajat Gupta)
 * [271a4d9] Fixed issue #521 Audit UI: Auditlog filters which doesn't require field valu.. (Rajat Gupta)
 * [c3ee2b0] Fix for refresh button will select all newly added file system automatically.  (Rajat Gupta)
 * [b27aba5] Fixed Audit Logs: unable to open the value input box for 'Event Start' and '.. (Rajat Gupta)
 * [93f4ebb] Fixing copyright date                                                          (Daniel Henrique Barboza)
 * [f43b9ca] Issue #531 Removed refresh of networkdevices on cancellation.                  (Megha Smriti)
 * [6fd2c63] Issue #525. Disabling Set OSA Port option for bond/vlan.                       (Megha Smriti)
 * [e73e078] Bug fix #528                                                                   (Megha Smriti)
 * [0711217] Remove role_key parameter                                                      (Aline Manera)
 * [f015905] Fixing copyright year of UI files after previous commit                        (Daniel Henrique Barboza)
 * [905d098] Bug fix #517: Restore system backup                                            (Bianca Carvalho)
 * [62f1414] Minor issue fixes in create partition                                          (Suresh Babu Angadi)
 * [9b75863] Fixing copyright year of storage_devs.py                                       (Daniel Henrique Barboza)
 * [a3b8cdb] Fix for ginger issue #526                                                      (Megha Smriti)
 * [d6c7a2f] Updating ginger.css with latest changes                                        (Daniel Henrique Barboza)
 * [efa56f9] Remove related functions to VGs Actions                                        (Ramon Medeiros)

ginger.py
#
# Project Ginger
#
# Copyright IBM Corp, 2014-2017
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA

import json
import os

from i18n import messages
from model import GingerModel
from wok import config
from wok.config import PluginConfig, PluginPaths
from wok.plugins.ginger.control import sub_nodes
from wok.root import WokRoot


class Ginger(WokRoot):
    def __init__(self, wok_options=None):
        object_store = config.get_object_store()
        objstore_dir = os.path.dirname(os.path.abspath(object_store))
        if not os.path.isdir(objstore_dir):
            os.makedirs(objstore_dir)

        self.model = GingerModel()
        super(Ginger, self).__init__(self.model)

        for ident, node in sub_nodes.items():
            setattr(self, ident, node(self.model))

        self.api_schema = json.load(open(os.path.join(os.path.dirname(
                                    os.path.abspath(__file__)), 'API.json')))
        self.domain = "ginger"
        self.depends = ['gingerbase']
        self.messages = messages
        self.paths = PluginPaths('ginger')

    def get_custom_conf(self):
        custom_config = PluginConfig('ginger')
        custom_config.update({
            '/help': {
                'tools.staticdir.on': True,
                'tools.nocache.on': True,
                'tools.staticdir.dir':  os.path.join(self.paths.ui_dir,
                                                     'pages/help')
            }
        })

        return custom_config
